phases/ephemeral/witx/wasi_ephemeral_sock.witx

```diff
@@ -11,34 +11,241 @@
   ;;; Linear memory to be accessed by WASI functions that need it.
   (import "memory" (memory))
 
+  ;;; Returns the local address to which the socket is bound.
+  ;;;
+  ;;; Note: This is similar to `getsockname` in POSIX
+  ;;;
+  ;;; When successful, the contents of the output buffer consist of an IP address,
+  ;;; either IP4 or IP6.
+  (@interface func (export "addr_local")
+    ;;; Host to resolve
+    (param $fd $fd)
+    ;;; The buffer where IP addresses will be stored
+    (param $buf (@witx pointer u8))
+    (param $buf_len $size)
+    (result $error $errno)
+  )
+
+  ;;; Returns the remote address to which the socket is connected to.
+  ;;;
+  ;;; Note: This is similar to `getpeername` in POSIX
+  ;;;
+  ;;; When successful, the contents of the output buffer consist of an IP address,
+  ;;; either IP4 or IP6.
+  (@interface func (export "addr_remote")
+    ;;; Host to resolve
+    (param $fd $fd)
+    ;;; The buffer where IP addresses will be stored
+    (param $buf (@witx pointer u8))
+    (param $buf_len $size)
+    (result $error $errno)
+  )
+
+  ;;; Open a socket
+  ;;;
+  ;;; The first argument to this function is a handle to an
+  ;;; address pool. The address pool determines what actions can
+  ;;; be performed and at which addresses they can be performed to.
+  ;;;
+  ;;; The address pool cannot be re-assigned. You will need to close
+  ;;; the socket and open a new one to use a different address pool.
+  ;;;
+  ;;; Note: This is similar to `socket` in POSIX using PF_INET
+  (@interface func (export "open")
+    ;;; Address pool file descriptor
+    (param $pool $fd)
+    ;;; Address family
+    (param $af $address_family)
+    ;;; Socket type, either datagram or stream
+    (param $socktype $sock_type)
+    (result $error $errno)
+    ;;; The opened socket
+    (result $fd $fd)
+  )
+
+  ;;; Close a socket (this is an alias for `fd_close`)
+  ;;; Note: This is similar to `close` in POSIX.
+  (@interface func (export "close")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    (result $error $errno)
+  )
```

Comment:
Is it desirable to have file descriptors be used interchangeably between different APIs? For example: should a socket fd be closeable by the `close` function of the filesystem API and vice-versa?

Reply:
Right now, it is desirable, because WASI is closely following POSIX in this respect. You can `close(2)` a socket, so the same should work in WASI.

In the future, when the [abstract types](https://github.com/WebAssembly/proposal-type-imports/blob/master/proposals/type-imports/Overview.md) mechanism in wasm is more developed, I expect we'll switch to making more type distinctions between different kinds of handles, and only provide the `close(2)` abstraction at the libc level.